tracking consent

mraveux opened the pull request:
This Pr add the TrackingConsent component, used to setup matomo tracking easily.

You can try it locally in the wallet, branch `matheo/matomo_tracking`, or using "Curd's Fancy Deployment Selector".

mraveux committed:
TrackingConsent component base / style

mraveux committed:
TrackingConsent: basic version / untested

mraveux committed:
TrackingConsent: turn public into static

- turn public methods into static ones
- turn "domain" prop into static const

danimoh commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    private denyConsent(): void {
+        this._setConsent({ allowsUsageData: false, allowsBrowserData: false });
```
Would be good to also disable tracking here in case it was already active.
I guess `_paq.push(['optUserOut']);` as described [here](https://developer.matomo.org/guides/tracking-javascript-guide#optional-creating-a-custom-opt-out-form) is suitable for this.
`_paq.push(['forgetUserOptOut']);` would then be needed to be called when we want to enable tracking again.

danimoh commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    private _initMatomo() {
+        // initialize matomo
```
If Matomo is already initialized, should not run this method again.
This is relevant if there are multiple instances of `TrackingConsent`

danimoh commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,309 @@
+            TrackingConsent._storage = JSON.parse(localStoredConsent);
+            TrackingConsent._setCookie(TrackingConsent.STORAGE_KEYS.MAIN, localStoredConsent);
+            localStorage.removeItem(TrackingConsent.STORAGE_KEYS.MAIN);
+            localStorage.removeItem(TrackingConsent.STORAGE_KEYS.SECOND);
```
This should be done in `created`, not as a side effect in a getter.

danimoh commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,309 @@
+        id="tracking-consent"
+        class="tracking-consent-banner nq-shadow"
+        :class="theme"
+        v-if="!consentKnown"
```
In some cases we want to have additional control over whether the ui should be displayed or not.
For example, when the multi currency checkout disclaimer is visible as overlay, the tracking consent overlay should not be rendered.
Can you add a `allowUi` prop for that purpose?

Yukiioz replied:
yeah sure

danimoh reviewed (changes requested):
See comments.

Yukiioz commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,309 @@
+        id="tracking-consent"
+        class="tracking-consent-banner nq-shadow"
+        :class="theme"
+        v-if="!consentKnown"
```
Sure, no problem.

mraveux committed:
TrackingConsent: safari hacks for browser bottom bar

mraveux committed:
TrackingConsent: PR requested changes - Multiple fix and improvements

- PR requested changes
- Listening for tab / window focus in case the user allows tracking on another tab / window or change browser settings / vpn

ToDo:
- Should use BottomOverlay
- Component doc comment

mraveux committed:
TrackingConsent: Add i18n

mraveux committed:
TrackingConsent: now centered on the bottom of the page

mraveux committed:
TrackingConsent: remove unnecessary ios fixes

mraveux committed:
TrackingConsent: debug with dashboard & fixes

- types fix
- setTrackerUrl & trackingScriptUrl urls fix
- add default setTrackerUrl value
- now using a window `focus` listener instead of document `visibilitychange` one

mraveux committed:
TrackingConsent: add some comments & multiple minor fixes

mraveux committed:
TrackingConsent: add execFunction method

mraveux committed:
TrackingConsent: add api reference link for options / now supporting null and string[] as options value’s type

mraveux committed:
TrackingConsent: some code cleaning

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+    @Prop({
+        type: Object,
+        default: () => ({
+            setSiteId: 1,
```
I would not set a default siteId. This is a _bad default_ and will lead to tracking on sites that don't change this. This prop should be `required` and the sideId should be a required setting.

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+
+    @Prop({
+        type: String,
+        default: () => TrackingConsent.DEFAULT_TRACKING_URL,
```
```suggestion
        default: () => TrackingConsent.DEFAULT_TRACKING_SCRIPT_URL,
```

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+        type: Boolean,
+        default: false,
+    })
+    public ignoreDoNotTrack: boolean;
```
Is it a good idea to make this a setting in the first place? Shouldn't we respect this setting?

Yukiioz replied:
We do respect this setting by default. But if someone want to not respect it, then he's free to do so. (see Daniel's comment https://github.com/nimiq/vue-components/pull/56#discussion_r334296302)

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+
+    @Prop({
+        type: String,
+        default: 'https://geoip.nimiq-network.com:8443/v1/locate',
```
Every other default setting is a constant, should this be, too?

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+            expirationDays: number,
+        },
+    ) {
+        const cookie = [cookieName + '=' + cookieValue];
+
+        if (options) {
+            if (!options.domain) options.domain = TrackingConsent.DEFAULT_COOKIE_DOMAIN;
+            if (!options.expirationDays) options.expirationDays = TrackingConsent.DEFAULT_COOKIE_EXPIRATION_DAYS;
+
+            cookie.push('domain=' + options.domain);
+            cookie.push('max-age=' + options.expirationDays * 24 * 60 * 60);
+        }
+
+        cookie.push('path=/');
+        document.cookie = cookie.join(';');
```
I believe this needs the `sameSite` setting as well nowadays.

Yukiioz replied:
`SameSite` require `Secure` sometimes, so i'll add both 👍

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+        DARK = 'dark',
+    }
+
+    export const LOCALSTORAGE_KEYS = [
+        'tracking-consent',
+        'tracking-consensus',
+    ];
```
Why two keys? Maybe add a comment to explain.

sisou commented on src/components/TrackingConsent.vue:
```diff
@@ -0,0 +1,467 @@
+    export const COOKIE_STORAGE_KEY = 'tracking-consent';
+
+    export const DEFAULT_COOKIE_DOMAIN = document.location.hostname;
+    export const DEFAULT_COOKIE_EXPIRATION_DAYS = 365 * 20;
```
20 years? :laughing:

Yukiioz replied:
Actually let me change it to 10 years, because 20 years won't work on 32bits systems.

mraveux committed:
TrackingConsent: Requested Changes

PR #56
Requested changes by Sisou
https://github.com/nimiq/vue-components/pull/56#pullrequestreview-455568171

- `options.setSiteId` doesn’t have a default value anymore, and is now required
- Renamed `DEFAULT_TRACKING_URL` to `DEFAULT_TRACKING_SCRIPT_URL`
- Moved the default `geoIpServer` value to a constant named `DEFAULT_GEOIP_SERVER_URL`
- Added possibility to set the cookie's attributes `secure` and `sameSite`
- Added a comment explaining the two `localStorage` keys
- Changed default lifetime of the cookie from 20 years to 10 years since 20 years would not work on 32bits systems. (the cookie would just disappear instantly)

mraveux committed:
TrackingConsent: Multiple changes & fixes

- now loading Matomo tracking script before checking for UI requirement, but disable cookie usage before user give consent (if in EEA)
- removed `allowBrowserData` since we’re tracking browser data by default, without asking for consent
- fixed `geoIpServer` @Prop default value
- now checking if `_paq` & `_mtm` are objects that have a `push` method, instead of checking if they’re Arrays (since they’re not Arrays, but objects with a single `push` method)
- removed the public static getter `allowsBrowserData` and the public static method `allowBrowserData `
- added a static public getter `isOutsideEEA`, since we can enable Cookies without consent for countries outside the EEA
- turned the `execFunction` method into a async one, allowing to `await` for the function result

mraveux committed:
TrackingConsent: trackPageView changes

- add a `trackPageView` public static method for easy page tracking on single-page apps
- set trackPageView by default to track first page view

mraveux committed:
TrackingConsent: if `geoIpServer` is set to `null`, the geoIp detection is disabled
